### ISSUE TITLE
box-default wrongly suggest on dyn trait 

### ISSUE THREAD

**Issue by flisky** [rust-lang/rust-clippy#9621]
### Summary

I have a function signature like `fn with_bar(self, bar: Box<dyn Bar>)`,  and `box-default` will suggest `Foo::new().with_bar(Box::default())` over `Foo::new().with_bar(Box::new(BarImpl::default))`, which clearly cannot be compiled.

### Lint Name

box-default

### Version

```text
rustc 1.66.0-nightly (8796e7a9c 2022-10-08)
binary: rustc
commit-hash: 8796e7a9cfd4c5c4f1de15ec1c53994ddf288665
commit-date: 2022-10-08
host: aarch64-apple-darwin
release: 1.66.0-nightly
LLVM version: 15.0.2
```

**Comment by llogiq** [comment]
@rustbot claim

I think it should still work with `Box::<BarImpl>::default()`. Can you check that it does?

**Comment by Vlad-Shcherbina** [comment]
`Box::<T>::default()` works, but Clippy suggests `Box::default()`, not `Box::<T>::default()`. The suggestion is incorrect.

Consider this artifical example:
```rust
trait MyTrait {}

#[derive(Default)]
struct MyStruct;

impl MyTrait for MyStruct {}

fn take_box_dyn(_: Box<dyn MyTrait>) {}

fn main() {
    let _: Box<dyn MyTrait> = Box::new(MyStruct::default());
    take_box_dyn(Box::new(MyStruct::default()));
}
```
Clippy says this:
```
warning: `Box::new(_)` of default value
  --> src\main.rs:11:31
   |
11 |     let _: Box<dyn MyTrait> = Box::new(MyStruct::default());
   |                               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: try: `Box::<MyStruct>::default()`
   |
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#box_default
   = note: `#[warn(clippy::box_default)]` on by default

warning: `Box::new(_)` of default value
  --> src\main.rs:12:18
   |
12 |     take_box_dyn(Box::new(MyStruct::default()));
   |                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: try: `Box::default()`
   |
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#box_default
```

So, for a variable of type `Box<dyn Trait>` it correctly suggests `Box::<T>::default()`, but for a function expecting `Box<dyn Trait>` it incorrectly and inconsistently suggests `Box::default()`.

```
> cargo clippy --version
clippy 0.1.68 (0468a00a 2022-12-17)
```

**Comment by rekby** [comment]
I have same problem: box to dyn trait and clippy suggest variant with Box::default(), which can't compiled.

**Comment by Eugeny** [comment]
Same on the latest nightly - this issue is only partially fixed.

**Comment by rwthompsonii** [comment]
I'm having an issue with a crate that provides a builder-pattern "default" function and clippy is complaining about it, even though the object does not (and cannot) implement Default.  I'm silencing the warning for now.

**Comment by llogiq** [comment]
@rwthompsonii can you provide code?

**Comment by rwthompsonii** [comment]
It's closed source, I'll try and get an unencumbered example tomorrow.
